app.go

```diff
@@ -404,6 +404,13 @@ func (app *App) provide(constructor interface{}) {
 		return
 	}
 
+	if a, ok := constructor.(Annotated); ok {
+		if err := app.container.Provide(a.Fn, dig.Name(a.Name)); err != nil {
```

Comment:
For the case where the constructor is just a function, we should probably
verify that we're not producing `fx.Annotated`s into the container. It would
be easy for someone to assume that they can do `func() fx.Annotated` to get
the same behavior so we can catch that now and give them a better error
message.

You should be able to adapt some of the code in `"internal/fxreflect".ReturnTypes`
to get a list of `reflect.Type`s being provided by the constructor and check
that one of them isn't `reflect.TypeOf(Annotated{})`. (The code in that
package has some duplication with dig; we plan on exposing APIs in dig to
reduce that in the future.)